src/module.c

```diff
@@ -60,6 +71,52 @@ int _RegisterDataTypes(RedisModuleCtx *ctx) {
 	return REDISMODULE_OK;
 }
 
+static void _PrepareModuleGlobals() {
+	assert(pthread_mutex_init(&_module_mutex, NULL) == 0);
+	graphs_in_keyspace = array_new(GraphContext *, 1);
+	process_is_child = false;
+}
+
+static void RG_ForkPrepare() {
+	/* At this point, a fork call has been issued. (We assume that this is because BGSave was called.)
+	 * Acquire the writer mutex of each graph to ensure that no locks are held, or else
+	 * the child process will deadlock when attempting to acquire that lock.
+	 * 1. If a writer thread is active, we'll wait until the writer finishes and releases the lock.
+	 * 2. Otherwise, no write in progress. Acquire the lock and release it immediately after forking. */
+
+	// Acquire the module-scoped lock; it will be released after forking.
+	assert(pthread_mutex_lock(&_module_mutex) == 0);
+
+	uint graph_count = array_len(graphs_in_keyspace);
+	for(uint i = 0; i < graph_count; i++) {
+		// Acquire each read-write lock as a reader to guarantee that no graph is being modified.
+		Graph_AcquireReadLock(graphs_in_keyspace[i]->g);
+	}
+
+	// Signal to the child that it is a forked process
+	process_is_child = true;
```

Comment:
Can't we use `RG_AfterForkChild` to set `process_is_child = true;`
Isn't `RG_AfterForkChild` being executed in the child process?

Reply:
👍 That is a much more sensible approach.